doc/encoding.rdoc

```diff
@@ -205,57 +205,74 @@ other than from the filesystem:
+from the stream are to be interpreted as characters.
 
 The default external encoding is:
 
 - UTF-8 for a text stream.
 - ASCII-8BIT for a binary stream.
 
-    f = File.open('t.rus', 'rb')
-    f.external_encoding  # => #<Encoding:ASCII-8BIT>
+The default external encoding is returned by method Encoding.default_external,
+and may be set by:
+
+- Ruby command-line options <tt>--external_encoding</tt> or <tt>-E</tt>.
+
+You can also set the default external encoding using method Encoding.default_external=,
+but doing so may cause problems; strings created before and after the change
+may have a different encodings.
 
-The external encoding may be set by the open option +external_encoding+:
+For an \IO or \File object, the external encoding may be set by:
 
-  f = File.open('t.txt', external_encoding: 'ASCII-8BIT')
-  f.external_encoding                               # => #<Encoding:ASCII-8BIT>
+- Open options +external_encoding+ or +encoding+, when the object is created;
+  see {Open Options}[rdoc-ref:IO@Open+Options].
 
-The external encoding may also set by method #set_encoding:
+For an \IO, \File, \ARGV, or \StringIO object, the external encoding may be set by:
 
-  f = File.open('t.txt')
-  f.set_encoding('ASCII-8BIT')
-  f.external_encoding  # => #<Encoding:ASCII-8BIT>
+- \Methods +set_encoding+ or (except for \StringIO) +set_encoding_by_bom+.
 
 ==== Internal \Encoding
 
-If not +nil+, the internal encoding specifies that the characters read
-from the stream are to be converted to characters in the internal encoding;
+The internal encoding, which is an \Encoding object or +nil+,
+specifies how characters read from the stream
+are to be converted to characters in the internal encoding;
 those characters become a string whose encoding is set to the internal encoding.
 
 The default internal encoding is +nil+ (no conversion).
-The internal encoding may set by the open option +internal_encoding+:
+It is returned by method Encoding.default_internal,
+and may be set by:
+
+- Ruby command-line options <tt>--internal_encoding</tt> or <tt>-E</tt>.
+
+You can also set the default internal encoding using method Encoding.default_internal=,
+but doing so may cause problems; strings created before and after the change
+may have a different encodings.
+
+For an \IO or \File object, the internal encoding may be set by:
 
-  f = File.open('t.txt', internal_encoding: 'ASCII-8BIT')
-  f.internal_encoding  # => #<Encoding:ASCII-8BIT>
+- Open options +internal_encoding+ or +encoding+, when the object is created;
+  see {Open Options}[rdoc-ref:IO@Open+Options].
 
-The internal encoding may also set by method #set_encoding:
+For an \IO, \File, \ARGV, or \StringIO object, the internal encoding may be set by:
 
-  f = File.open('t.txt')
-  f.set_encoding('UTF-8', 'ASCII-8BIT')
-  f.internal_encoding  # => #<Encoding:ASCII-8BIT>
+- \Methods +set_encoding+ or (except for \StringIO) +set_encoding_by_bom+.
```

Comment:
I think `set_encoding_by_bom` only changes the external encoding. E.g.

```
irb(main):012:0> str = "\xFE\xFF"
=> "\xFE\xFF"
irb(main):013:0> io = StringIO.new
=> #<StringIO:0x000000010a60d0a0>
irb(main):014:0> str = "\xFE\xFF"
=> "\xFE\xFF"
irb(main):015:0> io = StringIO.new(str)
=> #<StringIO:0x000000010a527410>
irb(main):016:0> io.external_encoding
=> #<Encoding:UTF-8>
irb(main):017:0> io.internal_encoding
=> nil
irb(main):018:0> io.set_encoding_by_bom
=> #<Encoding:UTF-16BE>
irb(main):019:0> io.external_encoding
=> #<Encoding:UTF-16BE>
irb(main):020:0> io.internal_encoding
=> nil
```

Reply:
Fixed.